internal/gps/checks.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"github.com/golang/dep/internal/gps/pkgtree"
+)
+
+func FindIneffectualConstraints(manifest Manifest, packageTree pkgtree.PackageTree) []ProjectRoot {
+	var stdLibFn func(string) bool
+	// will track return value here
+	ineffectuals := make([]ProjectRoot, 0, 10)
+
+	// flatten list of actual imports which should be checked against
+	reachmap, _ := packageTree.ToReachMap(true, true, false, nil /*ignores*/)
+	reach := reachmap.FlattenFn(stdLibFn)
+	imports := make(map[string]bool)
+	for _, imp := range reach {
+		imports[imp] = true
+	}
+
+	// if manifest is a RootManifest, use requires and ignores
+	if rootManifest := manifest.(RootManifest); rootManifest != nil {
+		// add required packages
+		for projectRoot, _ := range rootManifest.RequiredPackages() {
+			imports[projectRoot] = true
```

Comment:
it's probably worth checking to see if the user has entered any local packages names into the required list, as those would also be ineffectual rules.

That is, if `pkgtree.ImportRoot` is `github.com/foo/bar`, and there's a directive to require `github.com/foo/bar/baz`, and that package exists in the package tree, then we can indicate that it's an ineffectual require.